src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+        }
+
+        return result;
+    }
+
+    // Given a zero based offset, find the line text after it in the document
+    private subsequentText(offset: number,
+                           document: vscode.TextDocument): string {
+        const startPos = document.positionAt(offset);
+        // We don't know how long the line is so just return a really long one.
+        const endPos = startPos.translate(0, 1000);
+        return document.getText(new vscode.Range(startPos, endPos));
+    }
+
+    // Create a new token object with an appended scopeName
+    private addTokenScope(token,
+                          scopeName: string) {
+        // Only a shallow clone is required
+        const tokenClone = Object.assign({}, token);
+        tokenClone.scopes.push(scopeName);
+        return tokenClone;
+    }
+
+    // Given a list of grammar tokens, find the tokens that are comments and
+    // the comment text is either `# region` or `# endregion`.  Return a new list of tokens
+    // with custom scope names added, "custom.start.region" and "custom.end.region" respectively
+    private extractRegionScopeElements(tokens,
+                                       document: vscode.TextDocument) {
+        const result = [];
+
+        const emptyLine = new RegExp("^[\\s]+$");
+        const startRegionText = new RegExp("^#\\s*region\\b");
+        const endRegionText = new RegExp("^#\\s*endregion\\b");
+
+        tokens.forEach( (token, index) => {
+          if (token.scopes.includes("punctuation.definition.comment.powershell")) {
+            if (emptyLine.test(this.preceedingText(token.startIndex, document))) {
+              const commentText = this.subsequentText(token.startIndex, document);
+
+              if (startRegionText.test(commentText)) { result.push(this.addTokenScope(token, "custom.start.region")); }
+              if (endRegionText.test(commentText)) { result.push(this.addTokenScope(token, "custom.end.region")); }
+            }
+          }
+        });
+        return result;
+    }
+
+    // Given a list of tokens, return a list of matched tokens/line numbers regions
+    private matchGrammarTokens(tokens, document: vscode.TextDocument): IMatchedTokenList {
+        const matchedTokens = [];
+
+        // Find matching Braces   { -> }
+        this.matchScopeElements(tokens,
+                                "punctuation.section.braces.begin.powershell",
+                                "punctuation.section.braces.end.powershell",
+                                MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching brackets   ( -> )
+        this.matchScopeElements(tokens,
+                                "punctuation.section.group.begin.powershell",
+                                "punctuation.section.group.end.powershell",
+                                MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
```

Comment:
The `x`s used here and below should be renamed more descriptively.

Reply:
Fixed.